python/lsst/pipe/base/graphBuilder.py

```diff
@@ -507,58 +518,58 @@ def connectDataIds(self, registry, collections, userQuery):
-            for task in self.tasks:
-                quantumDataId = commonDataId.subset(task.dimensions)
-                quantum = task.quanta.get(quantumDataId)
-                if quantum is None:
-                    quantum = _QuantumScaffolding(task=task, dataId=quantumDataId)
-                    task.quanta[quantumDataId] = quantum
-                # Whether this is a new quantum or an existing one, we can now
-                # associate the DatasetRefs for this row with it.  The fact
-                # the fact that a Quantum data ID and a dataset data ID both
-                # came from the same result row is what tells us they should
-                # be associated.
-                # Many of these associates will be duplicates (because another
-                # query row that differed from this one only in irrelevant
-                # dimensions already added them), and we use sets to skip.
-                for datasetType in task.inputs:
-                    ref = refsForRow[datasetType.name]
-                    quantum.inputs[datasetType.name][ref.dataId] = ref
-                for datasetType in task.outputs:
-                    ref = refsForRow[datasetType.name]
-                    quantum.outputs[datasetType.name][ref.dataId] = ref
-        if n >= 0:
-            _LOG.debug("Finished processing %d rows from data ID query.", n+1)
-        else:
-            _LOG.debug("Received no rows from data ID query.")
-
-    def resolveDatasetRefs(self, registry, collections, run, *, skipExisting=True):
+        _LOG.debug("Submitting data ID query and materializing results.")
+        with registry.queryDataIds(self.dimensions,
+                                   datasets=list(self.inputs),
+                                   collections=collections,
+                                   where=userQuery,
+                                   ).materialize() as commonDataIds:
+            _LOG.debug("Expanding data IDs.")
+            commonDataIds = commonDataIds.expanded()
+            _LOG.debug("Iterating over query results to associate quanta with datasets.")
+            # Iterate over query results, populating data IDs for datasets and
+            # quanta and then connecting them to each other.
+            n = 0
+            for n, commonDataId in enumerate(commonDataIds):
+                # Create DatasetRefs for all DatasetTypes from this result row,
+                # noting that we might have created some already.
+                # We remember both those that already existed and those that we
+                # create now.
+                refsForRow = {}
+                for datasetType, refs in itertools.chain(self.inputs.items(), self.intermediates.items(),
+                                                         self.outputs.items()):
+                    datasetDataId = commonDataId.subset(datasetType.dimensions)
+                    ref = refs.get(datasetDataId)
```

Comment:
I'm using `get` here to avoid `try...__getitem__...except` for control flow, not to avoid a double lookup.  Or maybe I'm just not seeing what you're proposing as an alternative?